<?xml version="1.0" encoding="utf-8"?>
<search>
  
  
  
  <entry>
    <title>about Spring transaction</title>
    <link href="/blog/2020/12/05/%E5%85%B3%E4%BA%8Espring%E4%BA%8B%E5%8A%A1/"/>
    <url>/blog/2020/12/05/%E5%85%B3%E4%BA%8Espring%E4%BA%8B%E5%8A%A1/</url>
    
    <content type="html"><![CDATA[<h1 id="关于Spring事务"><a href="#关于Spring事务" class="headerlink" title="关于Spring事务"></a>关于Spring事务</h1><p><a href="https://my.oschina.net/u/4579410/blog/4525713">https://my.oschina.net/u/4579410/blog/4525713</a></p><p>看完这个明白的</p><p>想想aop，</p><p>犯的错：</p><blockquote><p>在一个serviceA的非事务方法中调用 this.func2()   func2是个事务方法，</p><p>func2中调用了 serviceB的func1()  serviceB.func1() 是事务方法</p></blockquote><blockquote><p>结果：    在serviceB.func1()最后抛异常， serviceB回滚了  serviceA.func2()的数据落库了。。</p></blockquote><p>原理 ：</p><p>在serviceA的代理对象的调用过程中，serviceA的方法内部调用了本service的方法，aop只在 最外层的方法（serviceA.func1）进行了增强，也就是 事务开启和回滚都应该加在这里，但是serviceA.func1 没有@Transaction。。。所以 serviceA.func2() 的声明式事务并没有生效。</p>]]></content>
    
    
    
  </entry>
  
  
  
  
</search>
